_source/_posts/2022-09-26-java-jetty-oauth.md

```diff
@@ -0,0 +1,1017 @@
+
+## Build a web service with Gradle and Jetty
+
+Next, you're going to look at how you can accomplish the same thing you did above, but this time with Gradle instead of Maven. Why? Because some people are allergic to XML and love DSLs instead.
+
+The Gradle-based app is located in the `gradle-jetty` subdirectory.
+
+[Gradle](https://docs.gradle.org/current/userguide/userguide.html) is a Groovy-based build system that uses a custom DSL (domain-specific language) for its build files. Because the build script files are essentially Groovy scripts, the build system is incredibly powerful, thus, very flexible.
+
+To run a Jetty server using Gradle, you'll use [the Gretty plugin](https://github.com/gretty-gradle-plugin/gretty).
+
+From the Gretty docs:
+
+> Gretty is a feature-rich Gradle plugin for running web apps on embedded servlet containers. It supports Jetty versions 7, 8, and 9, Tomcat versions 7 and 8, multiple web apps, and many more. It wraps servlet  container functions as convenient Gradle tasks and configuration DSL
+
+This project uses Gretty plugin version 4.0.3, which uses Jetty version 11.0.11. Since version 4, Gretty requires Jakarta over Javax. You can use Gretty 3 if you want to support non-Jakarta apps (this is explained on the Gretty GitHub page above).
+
+The two Java servlet classes are the same in both projects. I did find that Gretty and Gradle crashed unless I added an empty `src/main/webapp` directory (this wasn't necessary with Maven).
+
+The `build.gradle` file is analogous to the `pom.xml` file.
+
+```bash
+plugins {
+    id 'java'
+    id 'war'
+    id 'org.gretty' version '4.0.3'
+}
+
+repositories {
+    jcenter()
+}
+
+dependencies {
+    providedCompile 'jakarta.servlet:jakarta.servlet-api:6.0.0'
+}
+
+gretty {
+    contextPath = '/'
+}
+```
+
+This is about as simple as it gets. The same Jakarta-based servlet API is included as a provided dependency, and the Gretty plugin is included in the `plugins` block.
+
+The context path is configured to be the root path. If you want to see the other configuration properties, take a look at [the Gretty configuration docs](https://gretty-gradle-plugin.github.io/gretty-doc/Gretty-configuration.html).
+
+You can also look at the Gradle tasks that the plugin adds [in the docs](https://gretty-gradle-plugin.github.io/gretty-doc/Gretty-tasks.html).
+
+To run the web app, use the `appRun` task.
+
+```bash
+./gradlew appRun
+```
```

Comment:
Looks like it doesn't work with Java 19:

```
Unsupported class file major version 63
```

Java 17 either:

```
Unsupported class file major version 61
```

It does work with 11.

Reply:
@mraible Was this in reference to something specific?